mountpoint-s3/src/data_cache/disk_data_cache.rs

```diff
@@ -0,0 +1,312 @@
+//! Module for the on-disk data cache implementation.
+
+use std::fs;
+use std::io::{ErrorKind, Write};
+use std::ops::RangeBounds;
+use std::path::PathBuf;
+
+use base64ct::{Base64Url, Encoding};
+use bytes::{BufMut, BytesMut};
+use mountpoint_s3_crt::checksums::crc32c;
+use tracing::{error, trace, warn};
+
+use crate::data_cache::DataCacheError;
+
+use super::{BlockIndex, CacheKey, ChecksummedBytes, DataCache, DataCacheResult};
+
+/// On-disk implementation of [DataCache].
+///
+/// TODO: Store checksums on disk, reconstruct as [ChecksummedBytes] using same checksum avoiding recomputation.
+///
+/// TODO: Store additional metadata with each block such as expected S3 key, ETag, etc..
+pub struct DiskDataCache {
+    block_size: u64,
+    cache_directory: PathBuf,
+}
+
+impl DiskDataCache {
+    /// Create a new instance of an [DiskDataCache] with the specified `block_size`.
+    pub fn new(cache_directory: PathBuf, block_size: u64) -> Self {
+        DiskDataCache {
+            block_size,
+            cache_directory,
+        }
+    }
+
+    /// Get the relative path for the given block.
+    fn get_path_for_key(&self, cache_key: &CacheKey) -> PathBuf {
+        let mut path = self.cache_directory.join("v1");
```

Comment:
Besides/instead of the version (which is not really needed for a temporary cache), should we include the process id, or something equivalent, to ensure the path is unique?

Maybe for this type it could just be a token to be passed at construction.

Reply:
I'll push an update to this PR now, but not address this today. Let me think about it.

I left in the versioning while I'm not super happy about a) not purging the cache on mount and b) not having things like etag, s3key in the blocks themselves on disk.

Reply:
Versioning here seems good. We definitely need some safety around (a) not using a non-empty cache directory where we might overwrite existing files, and (b) coordinating enough to avoid multiple Mountpoints trying to use the same directory. But happy for that to come in a followup.